### PR TITLE
[Feature] Bookmarks Improvement

### PR DESCRIPTION
**Pull request by 4shutosh**
Task/Issue URL: 
#1438
#1439 
#1440

PR related to the feature proposal mentioned in discussion #1442 

### Description
- Users can create a bookmark folder from the edit bookmark screen.
-  Edit Bookmark shows the correct Bookmark Location, From `BrowserFragment`

### Steps to test this PR
1. Add a bookmark
2. Edit the bookmark from the menu or from the snackbar
3. Select the Location
4. New bookmark folder can be created.
5. After creating a new bookmark folder, select the bookmark folder
6. Save : bookmark saved inside the new bookmark folder.

### CODE REVIEW

**Comment by 4shutosh** [comment]
Screenshot_20210922-124907_DuckDuckGo
Screenshot_20210922-124913_DuckDuckGo
Screenshot_20210922-124944_DuckDuckGo

**Comment by cmonfortep** [comment]
@4shutosh thanks for your time here. As I posted in the different issues you created, let's first have a discussion before we introduce any improvement.

We accept contributions for bug fixes after we triage the issue and we add [the label "help wanted"](https://github.com/duckduckgo/Android/blob/develop/CONTRIBUTING.md#what-to-expect).

For feature requests/improvements, you need to open a discussion first. Inside the discussion, we will have a nice chat to know more about the proposal, understand what's the user value, etc. That's because we will need to share the proposed change with our Product team and/or Design team. If it's aligned with our roadmap and we are happy with the idea then we will be able to give you support with the improvements.

Hope that's ok with you, let's close this PR for now. We can open it again once I get the internal approval. Thanks again, and I think the proposal makes sense, but we need to check internally first 👍

**Comment by 4shutosh** [comment]
Thank you for the explanation @cmonfortep 🙌🏻 

That's unfortunate but will surely do the required. 

I've created a discussion here #1442

**Comment by 4shutosh** [comment]
Behavioral changes introduced in this PR:
- The edit bookmark shows the folder name now
- We can now create a new bookmark folder while editing the bookmark from the edit option either from snackbar or from the menu options.

@cmonfortep please review.
Do let me know with any changes are required along with new tests to be written or not.

**Comment by cmonfortep** [comment]
Thanks @4shutosh. 

"The edit bookmark shows the folder name now". Sorry if it was not clear, but we discarded de idea of showing folder names in the snackbar (see my comment: https://github.com/duckduckgo/Android/discussions/1442#discussioncomment-1551699)

**Comment by 4shutosh** [comment]
Sorry for the confusion. Yes it is not on the snackbar but on the edit saved site screen.
image

**Comment by 4shutosh** [comment]
Here the "Edited" is the bookmark folder name.

**Comment by cmonfortep** [comment]
Thanks. I will try to review the PR this week. Could you update the description? (just to reflect what this PR is finally addressing since we discarded some initial proposals)

**Comment by cmonfortep** [comment]
@4shutosh just to give you some heads up. Testing this right now, looks great. Our designer is taking a look, should give me feedback on Monday.

**Comment by 4shutosh** [comment]
Thank you for the update 🙌🏻

**Comment by cmonfortep** [comment]
@4shutosh thank you for your patience.

Feedback from the design team was really positive, that's really cool! They only asked if we can take the user back to the edit screen when the user created the new folder (same as in chrome).

Feedback is **optional** and I haven't checked how complex it is, so let me know if you are interested (and if you have time, it's ok if you can't) or not. In case you want to contribute on that, you can do it here or in another PR.

**Comment by 4shutosh** [comment]
That's great!

I'm interested in working on the feedback.

But I'll be able to do that (or any test related progress of this PR: as some tests are failing) after a week.

Please proceed with the tests fix if this feature is of priority for upcoming release within a week. In that case, I'll be continuing the feedback work on a new PR.

**Comment by cmonfortep** [comment]
Sounds good, I will review this PR as it is now, and further improvements can be done on a separated PR.

File: app/src/main/java/com/duckduckgo/app/bookmarks/ui/EditSavedSiteDialogFragment.kt
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/bookmarks/ui/EditSavedSiteDialogFragment.kt:33]: I would try to avoid creating 2 methods here if it's not really necessary.

Checking the implementations of `onSavedSiteBookmarkEdited` only 1 of the 3 is using bookmarkFolder -> `BrowserTabViewModel`. I will leave a comment there.

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:1576]: I don't think we need bookmarkFolder as parameter here. I would prefer to keep one single method and get benefit of `SavedSite` sealed class.

What about just receiving `bookmark`, and retrieve `bookmarkFolder` from db inside `editBookmark`?

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:996]: any reason method param should be nullable? if it's not a valid argument, I would remove nullability.

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:1594]: after updating the bookmark, I would retreive the bookmark folder using `getBookmarkFolder` here. That way we can remove the method param and go back to a single method + sealed class in other classes.

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:282]: Not sure about this approach: we are using a sealed class (`savedSite`) to represent a Bookmark/Favorite, but now we have a second param that only applies to Bookmarks (`bookmarkFolder`). If that's the case, it should be part of the sealed class as well, because it only applies to one of the types.

`SavedSite` is a domain (layer) sealed class.

In here, I think you are passing `bookmarkFolder` to show bookmark folder name in the dialog. If that's the case, my approach will be to create a sealed class specific for those dialog. Then you can map your savedSite into your new saved site and use it there. But please, double check this approach, I could be missing something.

**Review by cmonfortep — CHANGES_REQUESTED** [review]
Thanks, overall looks great. There are a few decisions we should double check, I left my comments bellow.

**Comment by 4shutosh** [comment]
Thank you for the detailed review @cmonfortep.
I'll understand and double check the issues within few days and make necessary ammendments.

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt
Comment (4shutosh) [reply app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:996]: instead of checking the nullability of the bookmark (of the saved site got from the bookmarksRepository) every time I use this function, I thought to add the null check inside the function, hence the parameter is nullable.

Will think of a better way.

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt
Comment (4shutosh) [reply app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:1576]: Sure, this can be done. 

Trying this right away.

File: app/src/main/java/com/duckduckgo/app/bookmarks/ui/EditSavedSiteDialogFragment.kt
Comment (4shutosh) [reply app/src/main/java/com/duckduckgo/app/bookmarks/ui/EditSavedSiteDialogFragment.kt:33]: Yes removing the bookmark folder from the parameter helped me remove this bad code.

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt
Comment (4shutosh) [reply app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:282]: You are right.
I've now used a separate data class for the change in saved site to process interaction with the views.

**Comment by 4shutosh** [comment]
@cmonfortep I've improved the areas of code mentioned by you. Please take a look.

Also, please help me with the falling tests. Can't find a meaningful message in the [log](https://app.bitrise.io/build/032fb40f-ad0b-42c5-8850-abf30c17cf89) for the failure of `virtual-device-testing-for-android`. 

I haven't written any new tests related to this PR, please let me know if any new tests are required due to the introduction of new behavior.

**Comment by cmonfortep** [comment]
Thanks, @4shutosh, will take a look next week. I'd suggest aligning with develop to ensure no conflicts/issues.

**Review by cmonfortep — CHANGES_REQUESTED** [review]
Thanks @4shutosh, haven't reviewed all the changes. I've only focused on the dialogs and the browsertabviewmodel. Left some comments bellow.

While testing this, I've noticed an unexpected behavior: After editing a bookmark (changing folder), a snackbar saying "bookmark added" appears. Testing our latest app, the behavior is different. We only show the snackbar when the user adds a new bookmark, not after editing the bookmark.

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt
Comment (4shutosh) [reply app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:200]: Yes you are right, the bookmarkFolder can be removed. Doing that right away.

File: app/src/main/java/com/duckduckgo/app/bookmarks/ui/SavedSiteDialogFragment.kt
Comment (4shutosh) [reply app/src/main/java/com/duckduckgo/app/bookmarks/ui/SavedSiteDialogFragment.kt]: Done

**Comment by 4shutosh** [comment]
Sorry for the delay. @cmonfortep
I've done the improvements in the code as mentioned by you. Also removed the unnecessary snackbar behaviour after edit of bookmark.

This is getting stretched more than expected. I'll try my best to complete the changes related to this PR as early as possible now. Please help me with the review. 

Thank you again. 🙌🏻

**Comment by malmstein** [comment]
@cmonfortep can you check on this?

File: app/src/main/java/com/duckduckgo/app/bookmarks/model/BookmarksRepository.kt
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/bookmarks/model/BookmarksRepository.kt:91]: Haven't checked, but I think filtering at SQL level is more efficient than getting the whole list of folders here, and then filtering. Can you create a new method in `BookmarkFoldersDao`, and just use it here?

File: app/src/main/java/com/duckduckgo/app/bookmarks/ui/EditSavedSiteDialogFragment.kt
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/bookmarks/ui/EditSavedSiteDialogFragment.kt:59]: `savedLocation.setText(it)` no need to call the same method twice.

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt:1385]: I want to check with someone from the team, what's the difference between `savedSiteChangedViewState.bookmarkFolder?.id` vs `savedSiteChangedViewState.savedSite.parentId` when we have a bookmark. So maybe I will ask to double check this part.

File: app/src/main/java/com/duckduckgo/app/bookmarks/ui/bookmarkfolders/BookmarkFoldersViewModel.kt
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/bookmarks/ui/bookmarkfolders/BookmarkFoldersViewModel.kt:74]: Test missing for this.

File: app/src/main/java/com/duckduckgo/app/bookmarks/ui/bookmarkfolders/BookmarkFoldersViewModel.kt
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/bookmarks/ui/bookmarkfolders/BookmarkFoldersViewModel.kt:44]: Sorry, can you explain a bit why we have added these 2 new properties to the viewstate? what are we trying to achieve?

**Review by cmonfortep — COMMENTED** [review]
Thanks and sorry for the delay in the review. As you said, this is taking longer than we expected and it was difficult to remember what was missing here.

I'm leaving some comments. So far it's looking good. Just something important to mention:
- be sure new code is properly tested if the class has a test (otherwise is not necessary, e.g: activities/fragments)
- when aligning with develop, you may get lot of conflicts. We have introduced changes in our code style, so expect some conflicts there.

Thanks.

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt
Comment (4shutosh) [reply app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:1642]: Yes, there is no need to update that. Did not realize we were using the flow directly from the repository.

File: app/src/main/java/com/duckduckgo/app/bookmarks/ui/bookmarkfolders/BookmarkFoldersViewModel.kt
Comment (4shutosh) [reply app/src/main/java/com/duckduckgo/app/bookmarks/ui/bookmarkfolders/BookmarkFoldersViewModel.kt:44]: In the process of understanding and making things work (was not able to understand the schema completely), primarily making use of the `getFlatFolderStructure` function these two properties were required.

But they are not required in the ViewState. I'll add them as local variables inside the viewModel.

Comment (4shutosh) [reply app/src/main/java/com/duckduckgo/app/bookmarks/ui/bookmarkfolders/BookmarkFoldersViewModel.kt:44]: I'm still not sure about the correctness of the implementation.

The major change in this class is the usage of `AddBookmarkFolderListener` to listen to the addition of a new bookmark folder and in turn update the view i.e the "flat folder structure".

As I was not able to understand what's happening in `getFlatFolderStructure` inside the `BookmarksDataRepository` I was only able to reuse the function.

To reuse the function there is a `rootFolderName` property required. Which was initially obtained from the string value (maybe to make translations compatible). So I did the same using a new Command object `NewFolderCreatedUpdateTheStructure`.

Comment (cmonfortep) [reply app/src/main/java/com/duckduckgo/app/bookmarks/ui/bookmarkfolders/BookmarkFoldersViewModel.kt:44]: @4shutosh I've been taking a look at this part. And I understand better what was your problem here.

This is my suggestion to simplify this:
- keep the `viewstate` as it was before. No need to add new properties.
- change the signature of `newFolderAdded` to receive those params from the activity.

Doing that you will simplify a lot this class.

File: app/src/main/java/com/duckduckgo/app/bookmarks/ui/bookmarkfolders/BookmarkFoldersActivity.kt
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/bookmarks/ui/bookmarkfolders/BookmarkFoldersActivity.kt:111]: This is my proposal to avoid storing those constants in the ViewModel. Just get it from intent.extras since they never change. 

``` 
   private fun setNewlyCreatedSelectedFolderResult() {
        viewModel.newFolderAdded(
                rootFolderName = getString(R.string.bookmarksSectionTitle),
                selectedFolderId = intent.extras?.getLong(KEY_BOOKMARK_FOLDER_ID) ?: 0,
                currentFolder = intent.extras?.getSerializable(KEY_CURRENT_FOLDER) as BookmarkFolder?
        )
    }
```

Comment (4shutosh) [reply app/src/main/java/com/duckduckgo/app/bookmarks/ui/bookmarkfolders/BookmarkFoldersActivity.kt:111]: Yes, this will make things simpler. Working on the conflict parts now will update the PR soon along with these implementations.

**Comment by 4shutosh** [comment]
> Feedback from the design team was really positive, that's really cool! They only asked if we can take the user back to the edit screen when the user created the new folder (same as in chrome).

I tried to work on this. This can be done in a couple of ways as per my understanding.

1 
```
    override fun onBookmarkFolderAdded(bookmarkFolder: BookmarkFolder, exitAndSetSelected: Boolean) {
        viewModelScope.launch(dispatcherProvider.io()) {
            bookmarksRepository.insert(bookmarkFolder)
        }
        if (exitAndSetSelected) onItemSelected(bookmarkFolder)
        else command.value = NewFolderCreatedUpdateTheStructure
    }
```
But this will not give us the ideal behavior and the list will be visible for a split second.

2. Removing the list screen (`Select Location`) from back stack and maybe implementing the interface of folder addition directly on the `Edit Bookmark` screen. But this is not possible in our case as the `Select Location` list screen i.e `BookmarkFoldersActivity` is an activity. All the screens inside the chrome bookmark flow would be Fragments.

Please take a look and let me know any other changes required.

**Comment by cmonfortep** [comment]
Feedback was optional. If you want to work on that, it's awesome! but my proposal is to do it on another PR once this one is merged. Just to avoid delaying more this feature.

File: app/src/main/java/com/duckduckgo/app/bookmarks/ui/bookmarkfolders/BookmarkFoldersViewModel.kt
Comment (4shutosh) [reply app/src/main/java/com/duckduckgo/app/bookmarks/ui/bookmarkfolders/BookmarkFoldersViewModel.kt]: Added please check

File: app/src/main/java/com/duckduckgo/app/bookmarks/model/BookmarksRepository.kt
Comment (4shutosh) [reply app/src/main/java/com/duckduckgo/app/bookmarks/model/BookmarksRepository.kt]: Created a new method please check.

**Comment by 4shutosh** [comment]
Sure I'll do that.
I've done all the changes required, resolved merge conflicts and the feature is working as expected. Please let me know any other changes required.

File: app/src/main/java/com/duckduckgo/app/bookmarks/model/BookmarksRepository.kt
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/bookmarks/model/BookmarksRepository.kt:103]: Can you add a test to cover this method?

What will happen when parentId is not found or doesn't return any folder? Please also test that scenario.

**Review by cmonfortep — COMMENTED** [review]
Thanks @4shutosh. We are almost done here 🎉  

I've just left one last comment to cover with a unit test the method in the repository. I've tested the functionality and it works as expected.

File: app/src/main/java/com/duckduckgo/app/bookmarks/model/BookmarksRepository.kt
Comment (4shutosh) [reply app/src/main/java/com/duckduckgo/app/bookmarks/model/BookmarksRepository.kt]: I've added the test.

In that case, the bookmark is treated as if it is stored in the default root folder under the name "Bookmarks".

**Review by cmonfortep — APPROVED** [review]
Thanks for the changes @4shutosh LGTM! 

Please, solve the conflicts with develop (probably the test has been moved into unit tests folder), and as soon as the check is green we will merge it. :)

**Comment by cmonfortep** [comment]
Sharing here the stacktrace of the failing test.
```
com.duckduckgo.app.bookmarks.model.BookmarksDataRepositoryTest > whenGetBookmarkFolderByParentIdThenReturnBookmarkFolderForParentId[test(AVD) - 5.1.1] FAILED 
	java.lang.AssertionError: expected:<BookmarkFolder(id=1, name=name, parentId=11, numBookmarks=0, numFolders=0)> but was:<null>
	at org.junit.Assert.fail(Assert.java:89)
```

**Comment by 4shutosh** [comment]
Yes, got the problem in the test. I've pushed a commit with a change in the test. The tests should pass now.

**Comment by 4shutosh** [comment]
Hahaha nooooo, another incoming conflict 😭. Merging again and pushing the same.

**Comment by 4shutosh** [comment]
Yayyy 🎉 thank you for your guidance @cmonfortep 🙌🏻

**Pull request merged**
